docs/dev/testing.rst

```diff
@@ -0,0 +1,306 @@
+Running pytest Directly
+-------------------------------------------------------------------------------
+Running tests directly with ``pytest`` gives you the ability to run only
+tests from a particular test module, or even a single individual test case.
+It's also faster because there's no testing environment to set up. Instead,
+it just uses your Python environment, which should be the ``pudl-dev`` conda
+environment discussed in :doc:`/dev/dev_setup`. This is convenient if you're
+debugging something specific, or developing new test cases, but it's not as
+robust as using Tox.
+
+Running specific tests
+^^^^^^^^^^^^^^^^^^^^^^
+To run the software unit tests with ``pytest`` directly:
+
+.. code-block:: console
+
+   $ pytest test/unit
+
+To run only the unit tests for the Excel spreadsheet extraction module:
+
+.. code-block:: console
+
+   $ pytest test/unit/extract/excel_test.py
+
+To run only the unit tests defined by a single test class within that module:
+
+.. code-block:: console
+
+   $ pytest test/unit/extract/excel_test.py::TestGenericExtractor
+
+Custom PUDL pytest flags
+^^^^^^^^^^^^^^^^^^^^^^^^
+We have defined several custom flags to control pytest's behavior when running
+the PUDL tests. You can always check to see what custom flags exist by running
+``pytest --help`` and looking at the ``custom options`` section:
+
+.. code-block:: console
+
+  custom options:
+  --live-dbs            Use existing PUDL/FERC1 DBs instead of creating temporary ones.
+  --tmp-data            Download fresh input data for use with this test run only.
+  --etl-settings=ETL_SETTINGS
+                        Path to a non-standard ETL settings file to use.
+  --gcs-cache-path=GCS_CACHE_PATH
+                        If set, use this GCS path as a datastore cache layer.
+  --sandbox             Use raw inputs from the Zenodo sandbox server.
+
+Run the output tests using a pre-existing PUDL database rather than building
+a new one for testing:
+
+.. code-block:: console
+
+  $ pytest --live-dbs test/integration/fast_output_test.py
+
+Run the ETL portion of the integration tests, and force it to download fresh
+input data to a temporary datastore:
+
+.. code-block:: console
+
+   $ pytest --tmp-data test/integration/etl_test.py
+
+Run the ETL portion of the integration tests, but using a non-standard settings
+file (equivalent to ``tox -e ferc1_solo``):
+
+.. code-block:: console
+
+  $ pytest --etl-settings=test/settings/ferc1-solo-test.yml test/integration/etl_test.py
+
+Check that the schemas of all historical FERC 1 databases are compatible with
+the most recent year, which we use to generate the schema for our FERC 1 SQLite
+database containing all years of FERC 1 data (equivalent to ``tox -e ferc1_schema``):
+
+.. code-block:: console
+
+  $ pytest --etl-settings test/settings/full-integration-test.yml test/integration/etl_test.py::test_ferc1_schema
+
+-------------------------------------------------------------------------------
+Data Validation
```

Comment:
should this come before the running pytest directly section? that feels more like a tangent compared to this

Reply:
Hmm. I dunno. I felt like the Data Validation was something folks were less like to want to run themselves since it entails multiple hours of database generation, and then test running. Whereas running individual tests seemed more useful in the context of someone trying to make a code contribution / debug a particular failure efficiently.

Reply:
yeah sure. i don't think it is a big deal either way

Reply:
Yeah on second thought I agree. Have rearranged.